powershell_completions.go

```diff
@@ -11,90 +8,253 @@ import (
+
+	# Prepare the command to request completions for the program.
+    # Split the command at the first space to seperate the programm and arguments
+    $Program,$Arguments = $Command.Split(" ",2)
+    $RequestComp="$Program %[2]s $Arguments"
+    __%[1]s_debug "RequestComp: $RequestComp"
+
+    # we cannont use $WordToComplete because it 
+    # has the wrong values if the curser was moved
+    # so use the last argument
+    if ($WordToComplete -ne "" ) {
+        $WordToComplete = $Arguments.Split(" ")[-1]
+    }
+    __%[1]s_debug "New WordToComplete: $WordToComplete"
+
+
+    $LastParam = $Command.Split(' ')[-1]
+    $LastChar = $LastParam.Split('')[-1]
+    $CountEqualSign = ($LastParam.Split('=')).count - 1
+    if ( $WordToComplete -eq "" -Or ($LastChar -eq "=" -And $CountEqualSign -eq 1 )) {
+        # If the last parameter is complete (there is a space following it)
+        # We add an extra empty parameter so we can indicate this to the go method.
+        __%[1]s_debug "Adding extra empty parameter"
+`+"        # We need to use `\"`\" to pass an empty argument a \"\" or '' does not work!!!"+`
+`+"        $RequestComp=\"$RequestComp\" + ' `\"`\"' "+`
+    }
+
+    # When completing a flag with an = (e.g., testprog -n=<TAB>)
+    # bash focuses on the part after the =, so we need to remove
+    # the flag part from $cur
+    #if [[ "${cur}" == -*=* ]]; then
+    #    flagPrefix="${cur%%=*}="
+    #    cur="${cur#*=}"
+    #fi
+
+    __%[1]s_debug "Calling $RequestComp"
+    #call the command store the output in $out and redirect stderr and stdout to null
+    # $Out is an array contains each line per element
+    Invoke-Expression -OutVariable out "$RequestComp" 2>&1 | Out-Null
+
+
+    # get directive from last line
+    [int]$Directive = $Out[-1].TrimStart(':')
+    if ($Directive -eq "") {
+        # There is not directive specified
+        $Directive = 0
+    }
+    __%[1]s_debug "The completion directive is: $Directive"
+    
+    # remove directive (last element) from out
+    $Out = $Out | Where-Object { $_ –ne $Out[-1] }
+    __%[1]s_debug "The completions are: $Out"
+
+    if (($Directive -band $ShellCompDirectiveError) -ne 0 ) {
+        # Error code.  No completion.
+        __%[1]s_debug "Received error from custom completion go code"
+        return
+    }
+
+    $Longest = 0
+    $Values = $Out | ForEach-Object {
+        #Split the output in name and description
+`+"        $Name, $Description = $_.Split(\"`t\",2)"+`
+        __%[1]s_debug "Name: $Name Description: $Description"
+
+        # Look for the longest completion so that we can format things nicely
+        if ($Longest -lt $Name.Length) {
+            $Longest = $Name.Length
+        }
+
+        # Set the description to a one space string if there is none set.
+        # This is needed because the CompletionResult does not accept an empty string as argument
+        if (-Not $Description) {
+            $Description = " "
```

Comment:
I notice that if there are no descriptions, we see a set of empty parentheses.  It is not so bad, but fish does not put them, but from the comment, I guess you don't have a choice?  For example:

powershell:
```
> ./testprog no<TAB>
nodesc  ( )                        nofile  (No file completion done)
```
fish:
```
> ./testprog no<TAB>
nodesc  nofile  (No file completion done)
```

Reply:
This is particularly visible when I use a `--no-descriptions` flag:
```
> ./testprog
cmdColonArgs                     ( )  help                             ( )
cmdLongDesc12345678901234567890  ( )  nodesc                           ( )
compAndFile                      ( )  nofile                           ( )
completion                       ( )  specialChar                      ( )
dirOnly                          ( )  subdir                           ( )
filterFile                       ( )  vargs                            ( )
fullFile                         ( )
```

Reply:
That should be no problem to fix

Reply:
Ok that's working now.

Reply:
Good!!